File: megacleaner/network/discovery.py
"""UDP discovery for MegaCleaner peers on the local network."""

from __future__ import annotations

import socket
import threading
import time
from collections.abc import Callable

from .protocol import (
    MSG_DISCOVER,
    MSG_HELLO,
    MSG_SOURCE,
    MSG_STOP_SOURCE,
    PORT,
    Peer,
    get_broadcast_address,
    get_hostname,
    get_local_ip,
    parse_message,
)


class DiscoveryService:
    def __init__(self, on_peer: Callable[[Peer], None] | None = None) -> None:
        self._on_peer = on_peer
        self._hostname = get_hostname()
        self._local_ip = get_local_ip()
        self._is_source = False
        self._http_port: int | None = None
        self._stop = threading.Event()
        self._thread: threading.Thread | None = None
        self._sock: socket.socket | None = None

    @property
    def local_ip(self) -> str:
        return self._local_ip

    @property
    def hostname(self) -> str:
        return self._hostname

    def set_source(self, http_port: int | None) -> None:
        self._is_source = http_port is not None
        self._http_port = http_port

    def start(self) -> None:
        if self._thread and self._thread.is_alive():
            return
        self._stop.clear()
        self._thread = threading.Thread(target=self._listen, daemon=True)
        self._thread.start()

    def stop(self) -> None:
        self._stop.set()
        if self._sock:
            try:
                self._sock.close()
            except OSError:
                pass

    def scan(self) -> None:
        self._send(MSG_DISCOVER)

    def announce_source(self, http_port: int) -> None:
        self.set_source(http_port)
        peer = Peer(self._hostname, self._local_ip, is_source=True, http_port=http_port)
        self._send(peer.to_source())

    def announce_stop_source(self) -> None:
        self.set_source(None)
        self._send(MSG_STOP_SOURCE)

    def _listen(self) -> None:
        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        sock.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
        sock.bind(("", PORT))
        sock.settimeout(0.5)
        self._sock = sock

        while not self._stop.is_set():
            try:
                data, addr = sock.recvfrom(4096)
            except OSError:
                if self._stop.is_set():
                    break
                continue

            msg_type, peer = parse_message(data)
            if not msg_type:
                continue

            if msg_type == MSG_DISCOVER:
                self._reply_hello(sock, addr[0])
            elif msg_type in (MSG_HELLO, MSG_SOURCE) and peer:
                if peer.ip != self._local_ip and self._on_peer:
                    self._on_peer(peer)

    def _reply_hello(self, sock: socket.socket, target_ip: str) -> None:
        peer = Peer(
            self._hostname,
            self._local_ip,
            is_source=self._is_source,
            http_port=self._http_port,
        )
        payload = peer.to_hello().encode("utf-8")
        try:
            sock.sendto(payload, (target_ip, PORT))
        except OSError:
            pass

    def _send(self, payload: str) -> None:
        message = payload.encode("utf-8")
        if "|" not in payload:
            message = f"MEGACLEANER|1|{payload}".encode("utf-8")

        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        sock.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
        try:
            sock.sendto(message, (get_broadcast_address(), PORT))
            sock.sendto(message, ("255.255.255.255", PORT))
        finally:
            sock.close()


def periodic_scan(service: DiscoveryService, interval: float, stop_event: threading.Event) -> None:
    while not stop_event.wait(interval):
        service.scan()

File: electron/discovery.js
const dgram = require("dgram");
const { DISCOVERY_SCAN_INTERVAL_MS, SOURCE_ANNOUNCE_INTERVAL_MS } = require("./constants");
const { sortPeers } = require("./peers");
const {
  PORT,
  MSG_DISCOVER,
  MSG_HELLO,
  MSG_SOURCE,
  MSG_SCREEN,
  MSG_SCREEN_STOP,
  getLocalIp,
  getHostname,
  isLocalIp,
  parseMessage,
  toHello,
  toSource,
  toScreen,
  toStopSource,
  toStopScreen,
  toDiscover,
  sendBroadcast,
} = require("./protocol");

class DiscoveryService {
  constructor(onPeersChanged) {
    this.onPeersChanged = onPeersChanged;
    this.hostname = getHostname();
    this.localIp = getLocalIp();
    this.isSource = false;
    this.httpPort = null;
    this.isStreaming = false;
    this.streamPort = null;
    this.classId = null;
    this.className = null;
    this.role = null;
    this.peers = new Map();
    this.socket = null;
    this.scanTimer = null;
    this.sourceAnnounceTimer = null;
    this.screenAnnounceTimer = null;
  }

  refreshNetworkIdentity() {
    this.hostname = getHostname();
    this.localIp = getLocalIp();
  }

  start() {
    this.socket = dgram.createSocket("udp4");
    this.socket.on("error", () => {});

    this.socket.on("message", (data, rinfo) => {
      const { type, peer } = parseMessage(data);
      if (!type) {
        return;
      }

      if (type === MSG_DISCOVER) {
        this.replyHello(rinfo.address);
        return;
      }

      if (
        (type === MSG_HELLO ||
          type === MSG_SOURCE ||
          type === MSG_SCREEN ||
          type === MSG_SCREEN_STOP) &&
        peer
      ) {
        if (isLocalIp(peer.ip)) {
          return;
        }
        this.upsertPeer(peer, type);
      }
    });

    this.socket.bind(PORT, () => {
      this.upsertPeer(this.getSelfPeer(), MSG_HELLO);
      this.scan();
      this.scanTimer = setInterval(() => this.scan(), DISCOVERY_SCAN_INTERVAL_MS);
    });
  }

  stop() {
    this.stopSourceAnnounceTimer();
    this.stopScreenAnnounceTimer();
    if (this.scanTimer) {
      clearInterval(this.scanTimer);
      this.scanTimer = null;
    }
    if (this.socket) {
      this.socket.close();
      this.socket = null;
    }
  }

  getSelfPeer() {
    return {
      hostname: this.hostname,
      ip: this.localIp,
      isSource: this.isSource,
      httpPort: this.httpPort,
      isStreaming: this.isStreaming,
      streamPort: this.streamPort,
      classId: this.classId,
      className: this.className,
      role: this.role,
      isSelf: true,
    };
  }

  setRoleContext(role) {
    this.role = role || null;
    this.upsertPeer(this.getSelfPeer(), MSG_HELLO);
  }

  setClassContext({ classId, className } = {}) {
    this.classId = classId || null;
    this.className = className || null;
    this.upsertPeer(this.getSelfPeer(), MSG_HELLO);
  }

  scan() {
    sendBroadcast(toDiscover());
  }

  announceSource(httpPort, classMeta = {}) {
    this.refreshNetworkIdentity();
    this.isSource = true;
    this.httpPort = httpPort;
    if (classMeta.classId) {
      this.classId = classMeta.classId;
      this.className = classMeta.className || null;
    }
    this.upsertPeer(this.getSelfPeer(), MSG_SOURCE);
    this.broadcastSource(httpPort);
    this.startSourceAnnounceTimer(httpPort);
  }

  broadcastSource(httpPort) {
    sendBroadcast(
      toSource({
        hostname: this.hostname,
        ip: this.localIp,
        httpPort: httpPort || this.httpPort,
        classId: this.classId,
        className: this.className,
      })
    );
  }

  startSourceAnnounceTimer(httpPort) {
    this.stopSourceAnnounceTimer();
    this.sourceAnnounceTimer = setInterval(() => {
      if (!this.isSource) {
        return;
      }
      this.refreshNetworkIdentity();
      this.upsertPeer(this.getSelfPeer(), MSG_SOURCE);
      this.broadcastSource(httpPort || this.httpPort);
    }, SOURCE_ANNOUNCE_INTERVAL_MS);
  }

  stopSourceAnnounceTimer() {
    if (this.sourceAnnounceTimer) {
      clearInterval(this.sourceAnnounceTimer);
      this.sourceAnnounceTimer = null;
    }
  }

  announceStopSource() {
    this.stopSourceAnnounceTimer();
    this.isSource = false;
    this.httpPort = null;
    this.upsertPeer(this.getSelfPeer(), MSG_HELLO);
    sendBroadcast(toStopSource());
  }

  announceScreen(streamPort) {
    this.refreshNetworkIdentity();
    this.isStreaming = true;
    this.streamPort = streamPort;
    this.upsertPeer(this.getSelfPeer(), MSG_SCREEN);
    this.broadcastScreen(streamPort);
    this.startScreenAnnounceTimer(streamPort);
  }

  broadcastScreen(streamPort) {
    sendBroadcast(
      toScreen({
        hostname: this.hostname,
        ip: this.localIp,
        streamPort: streamPort || this.streamPort,
      })
    );
  }

  startScreenAnnounceTimer(streamPort) {
    this.stopScreenAnnounceTimer();
    this.screenAnnounceTimer = setInterval(() => {
      if (!this.isStreaming) {
        return;
      }
      this.refreshNetworkIdentity();
      this.upsertPeer(this.getSelfPeer(), MSG_SCREEN);
      this.broadcastScreen(streamPort || this.streamPort);
    }, SOURCE_ANNOUNCE_INTERVAL_MS);
  }

  stopScreenAnnounceTimer() {
    if (this.screenAnnounceTimer) {
      clearInterval(this.screenAnnounceTimer);
      this.screenAnnounceTimer = null;
    }
  }

  announceStopScreen() {
    this.stopScreenAnnounceTimer();
    this.isStreaming = false;
    this.streamPort = null;
    this.upsertPeer(this.getSelfPeer(), MSG_SCREEN_STOP);
    sendBroadcast(toStopScreen());
  }

  replyHello(targetIp) {
    if (!this.socket) {
      return;
    }

    this.refreshNetworkIdentity();
    const payload = toHello(this.getSelfPeer());
    this.socket.send(Buffer.from(payload, "utf8"), PORT, targetIp, () => {});
  }

  upsertPeer(peer, messageType = MSG_HELLO) {
    const key = peer.ip;
    const existing = this.peers.get(key) || {};

    if (messageType === MSG_SOURCE) {
      peer = { ...existing, ...peer, isSource: true };
    } else if (messageType === MSG_SCREEN) {
      peer = { ...existing, ...peer, isStreaming: true };
    } else if (messageType === MSG_SCREEN_STOP) {
      peer = { ...existing, ...peer, isStreaming: false, streamPort: null };
    } else if (messageType === MSG_HELLO) {
      peer = {
        ...existing,
        ...peer,
        isSource: peer.isSource ?? existing.isSource ?? false,
        httpPort: peer.httpPort ?? existing.httpPort ?? null,
        isStreaming: Boolean(peer.isStreaming) || Boolean(existing.isStreaming),
        streamPort: peer.streamPort ?? existing.streamPort ?? null,
        classId: peer.classId ?? existing.classId ?? null,
        className: peer.className ?? existing.className ?? null,
        role: peer.role ?? existing.role ?? null,
      };
    }

    if (existing.isSelf) {
      peer = { ...peer, isSelf: true };
    }

    this.peers.set(key, peer);
    this.onPeersChanged(this.getPeerList());
  }

  getPeerList() {
    return sortPeers(Array.from(this.peers.values()));
  }
}

module.exports = { DiscoveryService };
